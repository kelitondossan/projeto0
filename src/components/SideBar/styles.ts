import styled from "styled-components";
interface ButtonProps {
    cartOpen: boolean;
  }

export const SideContainer = styled.aside<ButtonProps>`
    
    position: fixed;
    top: 0;
    bottom: 0;
    right: 0;
    padding: 20px 40px 120px;
    width: min(75vw, 580px);
    height: 100vh;
    background-color: var(--primary);
    box-shadow: -2px 2px 10px rgba(0, 0, 0, 0.25);
    opacity: ${props => (props.cartOpen ? "1" : "0")};
    visibility: ${props => (props.cartOpen ? "visible" : "hidden")};
    transform: ${props => (props.cartOpen ? "translateX(0%)" : "translateX(110%)")};
    transition: all 0.5s;

    
    @media (max-width: 768px) {
        padding: 20px;
        width: min(85vw, 480px);
    }
    
`;
export const SideHeader = styled.div`
    display: flex;
    color: #FFF;
    align-items: center;
    justify-content: space-between;
    div{
        width: 50%;
    }
    h2{
        font-size: 27px;
        font-weight: 700;
    }
    span{
        background: black;
        width: 38px;
        height: 38px;
        border-radius: 50%;
        display: flex;
        align-items: center;
        justify-content: center;
        font-size: 28px;
        cursor: pointer;
    }
    
    @media (max-width: 768px) {
        div{
            width: auto;
        }
}
`;
export const BottomButton = styled.button`
    position: absolute;
    display: flex;
    align-items: center;
    justify-content: center;
    border: none;
    bottom: 0;
    right: 0;
    width: 100%;
    height: 80px;
    font-size: 28px;
    font-weight: 700;
    background: #000;
    color: #fff;
    cursor: pointer;
`;

export const CartButton = styled.button`
    padding: 10px 20px;
    cursor: pointer;
    border-radius: 8px;
    font-size: 18px;
    border: none;
    display: flex;
    >span{
        margin-left: 8px;
    }
    
`;

export const SideProducts = styled.div`
    width: 100%;
    display: flex;
    align-items: center;
    flex-direction: column;
    padding: 10px;
    gap: 35px;
    margin-top: 30px;
    height: min(100vh, 500px);
    overflow: auto;
    ::-webkit-scrollbar{
        width: 2px;
        //display: none;
    }
    @media screen and (max-width: 1250px) {
        height: 400px;
    }
    @media (max-width: 768px) {
        height: 430px;
    }

`;

export const CartProducts = styled.div`
    display: flex;
    align-items: center;
    justify-content: space-between;
    padding: 20px 25px;
    background: #fff;
    width: 100%;
    border-radius: 8px;
    position: relative;
    
    
    .quantity-container{
        display: flex;
        align-items: center;
        gap: 15px;
        p{
            font-size: 6px;
        }
    }
    @media screen and (max-width: 768px) {
    flex-direction: column;
        .quantity-container{
            display: flex;
            width: 100%;
            flex-direction: row;
            align-items: center;
            justify-content: space-between;
            margin-top: 10px;
        }
    }
    .remove{
        position: absolute;
        top: -9px;
        right: -9px;
        z-index: 100;
        width: 18px;
        height: 18px;
        display: flex;
        align-items: center;
        justify-content: center;
        color: var(--white);
        background: black;
        border-radius: 50%;
        cursor: pointer;
        border: none;
    }
`;

export const ProductPrice = styled.span`
    font-size: 15px;
    font-weight: 700;

    @media (max-width: 768px) {
        padding: 10px;
        background-color: #373737;
        color: var(--white);
        border-radius: 5px;
    }
    
`;

export const ProductQuantity = styled.div`
    border-radius: 8px;
    border: 0.3px solid #bfbfbf;
    padding: 6px 6px;
    display: flex;
    align-items: center;
    gap: 10px;
    
    button{
        border: none;
        outline: none;
        background: var(--white);
        padding: 2px 8px;
        font-size: 14px;
        cursor: pointer;
    }
    .minor{
            border-right: 1px solid #bfbfbf;
        }
        .plus{
            border-left: 1px solid #bfbfbf;
        }
`;

export const TotalContainer = styled.div`
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin-top: auto;
    font-size: 28px;
    font-weight: 700;
    padding: 5px 0;
    color: #FFF;
    position: absolute;
    bottom:0;
    right: 0;
    margin-bottom: 90px;
    padding: 0 50px;
    @media screen and (max-width: 768px) {
        padding: 0 20px;
    }
`;