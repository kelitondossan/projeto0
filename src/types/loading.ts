interface LoadingProps {
    isloading: boolean
}

export default LoadingProps;