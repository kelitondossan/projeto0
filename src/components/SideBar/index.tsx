import React, { useState, useMemo, useEffect } from "react";
import { Product } from "@/types/products";
import {
  SideContainer,
  SideHeader,
  BottomButton,
  CartButton,
  SideProducts,
  CartProducts,
  ProductPrice,
  ProductQuantity,
  TotalContainer,
} from "./styles";
import { FaShoppingCart } from "react-icons/fa";

import { useDispatch, useSelector } from "react-redux";
import {
  increase,
  decrease,
  getCartTotal,
  initialState,
  clear,
} from "../../features/cartSlice";

interface ButtonProps {
  cartOpen: boolean;
}

const SideBar: React.FC<ButtonProps> = ({ cartOpen }) => {
  const dispatch = useDispatch();
  const { cartItems } = useSelector((state: initialState) => state.cart);
  const cart = useSelector((state: initialState) => state.cart);
  const [isOpen, setIsOpen] = useState(false);
  const handleCLick = () => setIsOpen(!isOpen);

  useEffect(() => {
    dispatch(getCartTotal(cart));
  }, [cart]);

  return (
    <>
      <CartButton onClick={handleCLick}>
        <FaShoppingCart />
        <span>{cart.totalQuantity.toString()}</span>
      </CartButton>
      <SideContainer cartOpen={isOpen}>
        <div className="content">
          <SideHeader>
            <div>
              <h2>Carrinho de compras</h2>
            </div>
            <span onClick={handleCLick}>x</span>
          </SideHeader>
          <div className="overflow-container">
            <SideProducts>
              {cartItems.map((product) => (
                <CartProducts key={product.id}>
                  <img src={product.photo} width={75} height={70} />
                  <span>{product.name}</span>
                  <div className="quantity-container">
                    <ProductQuantity>
                      <button
                        className="minor"
                        onClick={() => dispatch(decrease(product))}
                      >
                        -
                      </button>
                      <span>{product.quantity}</span>
                      <button
                        className="plus"
                        onClick={() => dispatch(increase(product))}
                      >
                        +
                      </button>
                    </ProductQuantity>

                    <ProductPrice>
                      R${product.price * product.quantity}
                    </ProductPrice>
                  </div>
                  <button
                    className="remove"
                    onClick={() => dispatch(clear(product.id))}
                  >
                    x
                  </button>
                </CartProducts>
              ))}
            </SideProducts>
          </div>

          <TotalContainer>
            <span>Total:</span>
            <span>R${cart.totalPrice}</span>
          </TotalContainer>
          <BottomButton>Finalizar Compra</BottomButton>
        </div>
      </SideContainer>
    </>
  );
};

export default SideBar;
