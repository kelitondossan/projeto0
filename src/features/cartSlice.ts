import { Product } from "@/types/products";
import { bindActionCreators, createSlice } from "@reduxjs/toolkit";
import { store } from "./store";
import { ShowToast } from "../components/ToastMessage";

type ProductState = {
    cartItems: Array<Product>,
    totalQuantity: Number,
    totalPrice: number
};

const initialState = {
    cartItems: [],
    totalPrice: 0,
    totalQuantity: 0
} as ProductState;

const CartSlice = createSlice({
    name: "cart",
    initialState,
    reducers:{
        addToCart: (state, action) => {
            const addItem = state.cartItems.find(
                (Cartitem) => Cartitem.id === action.payload.id
            );
            addItem ? (addItem.quantity = addItem.quantity + 1 ) : state.cartItems.push({...action.payload, quantity: 1});
            
            ShowToast({type: 'success', message: `${action.payload.name} No carrinho`})
        },
        increase: (state, action) => {
            
            const itemIndex = state.cartItems.findIndex(
                (cartItem) => cartItem.id === action.payload.id
            );
            //if (itemIndex === -1) return;
            state.cartItems[itemIndex].quantity++;
        },
        decrease: (state, action) =>{
            const itemIndex = state.cartItems.findIndex(
                (cartItem) => cartItem.id === action.payload.id
            );
            state.cartItems[itemIndex].quantity--;
            if (state.cartItems[itemIndex].quantity === 0) {
                state.cartItems.splice(itemIndex, 1)
            } else {
                state.cartItems[itemIndex].quantity < 1;
            };
        },
        removeToCart: (state, action) =>{
            state.cartItems =  state.cartItems.filter((cartItem) => cartItem.id !== action.payload);
        },

        getCartTotal: (state, action) => {
            let { total, quantity } = state.cartItems.reduce(
                (cartTotal, cartItem) =>{
                    const { price, quantity } = cartItem;
                    const itemTotal = price * quantity;

                    cartTotal.total += itemTotal;
                    cartTotal.quantity += quantity;

                    return cartTotal;
                },
                {
                    total: 0,
                    quantity: 0
                }
            );
            state.totalQuantity = quantity;
            state.totalPrice = total;
        },
        
        clear: (state, action) => {
            state.cartItems = state.cartItems.filter((cartItem) => cartItem.id !== action.payload);
        }
    }
})
export type initialState = ReturnType<typeof store.getState>
export const { addToCart, increase, decrease, removeToCart, clear, getCartTotal } = CartSlice.actions;
export default CartSlice.reducer