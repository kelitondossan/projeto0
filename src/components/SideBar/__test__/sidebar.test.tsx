import { Product } from "@/types/products";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../../../features/store";
import "@testing-library/jest-dom";
import SideBar from "..";

describe("SideBar component", () => {
  it("should render the cart component correctly", () => {
    render(
      <Provider store={store}>
        <SideBar cartOpen />
      </Provider>
    );
    expect(SideBar).toMatchSnapshot();
  });
});
