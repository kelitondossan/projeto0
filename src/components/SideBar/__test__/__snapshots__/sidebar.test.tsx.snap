// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBar component should render the cart component correctly 1`] = `[Function]`;
